### DISCUSSION TITLE
Accounting (as in Shorewall)

### DISCUSSION THREAD

**Discussion by Talkless** [FoobarOy/foomuuri#38] [Ideas]
Shorewall has neat feature: accounting definitions that allows monitoring in-out traffic of TCP/UDP ports you are interested in:
https://shorewall.org/Accounting.html

There's Munin plugin to fetch data and render graphs:
https://gallery.munin-monitoring.org/plugins/munin-contrib/shorewall-accounting_/

For example, one can monitor HTTP traffic:

```
# cat /etc/shorewall/accounting 
#ACTION                 CHAIN   SOURCE      DEST        PROTO   DPORT   SPORT   USER    MARK    IPSEC
http_s_in:COUNT         -       eth0        -           tcp     80
http_s_out:COUNT        -       -           eth0        tcp     -       80
```

Shorewall is deprecated (also it's Munin plugin runs also on deprecated Python2) and this feature seems missing in Foomuuri...

**Comment by kimheino** [comment]
Currently Foomuuri supports:

```
prerouting {
  tcp sport 80 443 iifname wlp0s20f3 counter  # incoming packets on client side, reverse for server side
}

postrouting {
  tcp dport 80 443 oifname wlp0s20f3 counter  # outgoing packets on client side
}
```

Foomuuri also supports named counters and listing them in pretty format, but apparently they don't work in pre/postrouting. I'll fix that.

Anyway, I would much rather configure "accounting" in zone-zone section. I'll have to think about this more.

**Comment by kimheino** [comment]
Commit c17dd40 fixes named counters in prerouting:

```
prerouting {
  tcp sport 22 counter ssh_in continue
  tcp sport 22 counter ssh_total continue
}

postrouting {
  tcp dport 22 counter ssh_out continue
  tcp dport 22 counter ssh_total continue
}
```

```
#> foomuuri list counter
table inet foomuuri {
	counter ssh_in {
		packets 15 bytes 2020
	}
	counter ssh_total {
		packets 35 bytes 3640
	}
	counter ssh_out {
		packets 20 bytes 1620
	}
}
```

**Comment by kimheino** [comment]
Commit bc63048 adds `-conntrack` flag. Your shorewall-example is something like this:

```
public-localhost {
  # count
  tcp dport 80 443 counter http_in continue -conntrack    # localhost is server
  udp dport 443 counter http_in continue -conntrack       # http/3 

  # accept
  http
  https
  ...
}

localhost-public {
  # count
  tcp sport 80 443 counter http_out continue -conntrack
  udp sport 443 counter http_out continue -conntrack

  # outgoing http replies will be accepted by conntrack/established
  ...
}
```

Is this and above pre/postrouting enough to count traffic? Both examples will see and count same traffic.

TODO: Munin plugin to graph named counters.

**Reply by Talkless** [reply]
Thanks @kimheino , that was fast!

> Is this and above pre/postrouting enough to count traffic?

I believe so. I'm not too big of expert to think of cases where this could be not enough.

**Comment by kimheino** [comment]
Munin plugin to graph counters: https://github.com/munin-monitoring/contrib/pull/1434

**Comment by Talkless** [comment]
Just amazing... thanks!
